addon/TelemetryRappor.jsm

```diff
@@ -296,66 +345,91 @@ function getRandomFloat() {
+  }
+  return secret;
+}
+
+/**
+ * Returns the cohort. If the cohort is not stored, it's randomly generated
+ * in the range [0, cohorts] and stored in the preferences.
+ * 
+ * @param {string} name - Name to store the cohort.
+ * @param {int} cohorts - Max number of cohorts.
+ */
+function getCohort(name, cohorts){
+  let cohort = null;
+  try {
+    cohort = Services.prefs.getIntPref(PREF_RAPPOR_PATH + name + ".cohort");
+  } catch (e) {
+    log.error("Error getting the cohort", e);
+  }
+  if (!cohort) {
+    cohort = Math.floor(getRandomFloat() * cohorts);
+    Services.prefs.setIntPref(PREF_RAPPOR_PATH + name + ".cohort", cohort);
+  }
+  return cohort;
 }
 
 var TelemetryRappor = {
   /**
    * Receives the parameters for RAPPOR and returns the Instantaneosu Randomized Response.
    * @param {string} name - Name of the experiment. Used to store the preferences.
    * @param {string} value v - Value to submit
+   * @param {constant} method  - Method used to encode. Supported: nsICryptoHash.MD5 and nsICryptoHash.SHA256
    * @param {Object} params - The parameters for the RAPPOR algorithm.
    * @param {integer} params.filterSize k - Size of the bloom filter in bytes.
    * @param {integer} params.numHashFunctions h - Number of hash functions.
    * @param {integer} params.cohorts m - Number of cohorts to use.
    * @param {float} params.f - Value for probability f.
    * @param {float} params.p - Value for probability p.
    * @param {float} params.q - Value for probability q.
+   * @param {boolean} isSimulation - Boolean that indicates if the addon is run in a simulation.
+   * @param {int} simulatedCohort - Cohort for the simulation.
    *
    * @return An object containing the cohort and the encoded value in hex.
    */
-  createReport(name, value, params) {
-    // Generate the RAPPOR secret. This secret never leaves the client.
-    let secret = null;
-    log.debug("HEYYY");
-    try {
-      secret = Services.prefs.getCharPref(PREF_RAPPOR_SECRET);
-      if (secret.length != 64) {
-        secret = null;
-      }
-    } catch (e) {
-      log.error("Error getting secret from prefs", e);
-    }
-    if (!secret) {
-      let randomArray = new Uint8Array(32);
-      crypto.getRandomValues(randomArray);
-      secret = bytesToHex(randomArray);
-      Services.prefs.setCharPref(PREF_RAPPOR_SECRET, secret);
-    }
-
-    // If we haven't self-selected a cohort yet for this measurement, then do so now,
-    // otherwise retrieve the cohort.
-    let cohort = null;
-    try {
-      cohort = Services.prefs.getIntPref(PREF_RAPPOR_PATH + name + ".cohort");
-    } catch (e) {
-      log.error("Error getting the cohort", e);
-    }
-    if (!cohort) {
-      cohort = Math.floor(getRandomFloat() * params.cohorts);
-      Services.prefs.setIntPref(PREF_RAPPOR_PATH + name + ".cohort", cohort);
-    }
-
+  createReport(name, value, params, method, isSimulation = false, simulatedCohort = null) {
```

Comment:
This doesn't really need to know about the simulation. Just make it accept an optional value for the `cohort`: use the provided one or generate randomly if it's not provided.

Reply:
Done